python/ray/autoscaler/updater.py

```diff
@@ -294,6 +295,66 @@ def remote_shell_command_str(self):
             self.ssh_private_key, self.ssh_user, self.ssh_ip)
 
 
+class DockerCommandRunner(SSHCommandRunner):
+    def __init__(self, docker_config, **common_args):
+        self.ssh_command_runner = SSHCommandRunner(**common_args)
+        self.docker_name = docker_config["container_name"]
+        self.docker_config = docker_config
+        self.home_dir = None
+
+    def run(self,
+            cmd,
+            timeout=120,
+            exit_on_fail=False,
+            port_forward=None,
+            with_output=False):
+
+        return self.ssh_command_runner.run(
+            cmd,
+            timeout=timeout,
+            exit_on_fail=exit_on_fail,
+            port_forward=None,
+            with_output=False)
+
+    def check_container_status(self):
+        no_exist = "not present"
+        cmd = get_docker_check(self.docker_name) + " ".join(
+            ["||", "echo", quote(no_exist)])
+        output = self.ssh_command_runner.run(
+            cmd, with_output=True).decode("utf-8").strip()
+        if no_exist in output:
+            return False
+        return output
+
+    def run_rsync_up(self, source, target):
+        self.ssh_command_runner.run_rsync_up(source, target)
+        if self.check_container_status():
+            self.ssh_command_runner.run("docker cp {} {}:{}".format(
+                target, self.docker_name, self.docker_expand_user(target)))
+
+    def run_rsync_down(self, source, target):
+        self.ssh_command_runner.run("docker cp {}:{} {}".format(
+            self.docker_name, self.docker_expand_user(source), source))
+        self.ssh_command_runner.run_rsync_down(source, target)
+
+    def remote_shell_command_str(self):
+        inner_str = self.ssh_command_runner.remote_shell_command_str().replace(
+            "ssh", "ssh -tt", 1).strip("\n")
+        return inner_str + " docker exec -it {} /bin/bash\n".format(
+            self.docker_name)
+
+    def docker_expand_user(self, string):
+        if string.find("~") == 0:
+            if self.home_dir is None:
+                self.home_dir = self.ssh_command_runner.run(
+                    "docker exec {} env | grep HOME | cut -d'=' -f2".format(
+                        self.docker_name),
+                    with_output=True).decode("utf-8").strip()
+            return string.replace("~", self.home_dir)
```

Comment:
I'm not sure if this is always going to work

Reply:
Is this because of `$HOME` changing over time or some other, more subtle reason. 
I can also change `string.replace(..)` to `string.replace(..., max_calls=)`.

Reply:
maybe even just take it out; I don't think docker supports tilde in the first place.

Reply:
That's the motivation for doing this--to provide the same result if a file is rsync'd to `~/dir/file` on a VM or a container.